csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -2213,374 +2240,564 @@ module Internal {
+            or
+            succSplit.hasEntry(pred, succ, c)
+          )
+        )
       )
     }
 
     /**
-     * Holds if `cfe` is a last element in the `finally` block belonging to
-     * `try` statement `try`, with completion `c`
+     * Holds if `pred` with splits `predSplits` can exit the enclosing callable
+     * `succ` with type `t`.
      */
-    private predicate isLastInFinally(TryStmt try, ControlFlowElement cfe, Completion c) {
-      cfe = getAFinallyDescendant(try) and
-      cfe = lastTryStmtFinally(try, c)
-    }
-
-    // Successor relation for `finally` to callable exit nodes
-    predicate succByTypeFinallyCallableExit(FinallyControlFlowElement pred, FinallySplitType predSplitType, ControlFlowEdgeType t, Callable succ) {
+    predicate succExitSplits(ControlFlowElement pred, Splits predSplits, Callable succ, ControlFlowEdgeType t) {
       exists(Completion c |
-        t.matchesCompletion(c) and
-        pred.isRelevantSplitType(predSplitType) and
-        pred = last(succ.getBody(), c) and
-        c.isValidCallableExitCompletion()
-        |
-        // No splitting: any completion inside the `finally` block is valid
-        isLastInFinally(_, pred, c) and
-        predSplitType instanceof ControlFlowEdgeSuccessor
-        or
-        // Splitting: completion must respect the split type
-        predSplitType.matchesCompletion(c)
-        or
-        // Ignore splitting: The `finally` block can itself complete abnormally
-        isLastInFinally(_, pred, c) and
-        not c instanceof NormalCompletion
+        t.matchesCompletion(c) |
+        succ = succExit(pred, c) and
+        predSplits.appliesTo(pred) and
+        forall(Split predSplit |
+          predSplit = predSplits.getASplit() |
+          predSplit.hasExit(pred, c)
+        )
       )
     }
+  }
+  import Splitting
 
-    // Successor relation for `finally` to non-`finally` nodes
-    predicate succByTypeFinallyNonFinally(FinallyControlFlowElement pred, FinallySplitType predSplitType, ControlFlowEdgeType t, NonFinallyControlFlowElement succ) {
-      exists(Completion c |
-        t.matchesCompletion(c) and
-        pred.isRelevantSplitType(predSplitType) and
-        succ = succ(pred, c)
-        |
-        // No splitting: any completion inside the `finally` block is valid
-        isLastInFinally(_, pred, c) and
-        predSplitType instanceof ControlFlowEdgeSuccessor
-        or
-        // Splitting: completion must respect the split type
-        predSplitType.matchesCompletion(c)
+  /** Provides logic for calculating reachable control flow nodes. */
+  module Reachability {
+    private predicate startsSplits(ControlFlowElement cfe) {
+      cfe = succEntry(_)
+      or
+      exists(Split s |
+        s.hasEntry(_, cfe, _)
         or
-        // Ignore splitting: The `finally` block can itself complete abnormally
-        isLastInFinally(_, pred, c) and
-        not c instanceof NormalCompletion
+        s.hasExit(_, cfe, _)
```

Comment:
I don't understand this line.

Reply:
I guess the `startsSplits()` predicate should be documented:
```
Holds if `cfe` is a control flow element where the set of possible
splits may be different from the set of possible
splits for one of `cfe`'s possible predecessors.
```
Does it make sense then?